### ISSUE TITLE
Panic when sending time.Time

### ISSUE THREAD

**Issue by heri16** [halturin/ergo#63]
Somewhere in this library will panic when sending a golang `time.Time`:

```
panic: reflect.Value.Interface: cannot return value obtained from unexported field or method
```

The golang reflect documentation suggests to use `reflect.Value.CanInterface()` before calling `reflect.Value.Interface()`. If !v.CanInterface(), we could use `reflect.Value.String()` instead or the unexported access technique in the below link.

- https://forum.golangbridge.org/t/panic-reflect-value-interface-cannot-return-value-obtained-from-unexported-field-or-method/17462

- https://stackoverflow.com/questions/42664837/access-unexported-fields-in-golang-reflect/43918797#43918797

- https://play.golang.org/p/IgjlQPYdKFR

**Comment by halturin** [comment]
May I ask you provide the way how it can be reproduced?

**Comment by heri16** [comment]
To reproduce, modified demoGenServer as below:

https://github.com/halturin/ergo/blob/master/examples/genserver/demoGenServer.go

```go
package main

import (
	"flag"
	"fmt"
	"time"

	"github.com/halturin/ergo"
	"github.com/halturin/ergo/etf"
)

// GenServer implementation structure
type demoGenServ struct {
	ergo.GenServer
	process *ergo.Process
}

type state struct {
	i int
}

var (
	GenServerName    string
	NodeName         string
	Cookie           string
	err              error
	ListenRangeBegin int
	ListenRangeEnd   int = 35000
	Listen           string
	ListenEPMD       int

	EnableRPC bool
)

// Init initializes process state using arbitrary arguments
// Init(...) -> state
func (dgs *demoGenServ) Init(p *ergo.Process, args ...interface{}) interface{} {
	fmt.Printf("Init: args %v \n", args)
	dgs.process = p
	return state{i: 12345}
}

// HandleCast serves incoming messages sending via gen_server:cast
// HandleCast -> ("noreply", state) - noreply
//		         ("stop", reason) - stop with reason
func (dgs *demoGenServ) HandleCast(message etf.Term, state interface{}) (string, interface{}) {
	fmt.Printf("HandleCast: %#v\n", message)
	switch message {
	case etf.Atom("stop"):
		return "stop", "they said"
	}
	return "noreply", state
}

// HandleCall serves incoming messages sending via gen_server:call
// HandleCall -> ("reply", message, state) - reply
//				 ("noreply", _, state) - noreply
//		         ("stop", reason, _) - normal stop
func (dgs *demoGenServ) HandleCall(from etf.Tuple, message etf.Term, state interface{}) (string, etf.Term, interface{}) {
	fmt.Printf("HandleCall: %#v, From: %#v\n", message, from)

	reply := etf.Term(etf.Tuple{etf.Atom("error"), etf.Atom("unknown_request")})
	switch message {
	case etf.Atom("time"):
		reply = time.Now()
	}
	return "reply", reply, state
}

// HandleInfo serves all another incoming messages (Pid ! message)
// HandleInfo -> ("noreply", state) - noreply
//		         ("stop", reason) - normal stop
func (dgs *demoGenServ) HandleInfo(message etf.Term, state interface{}) (string, interface{}) {
	fmt.Printf("HandleInfo: %#v\n", message)
	return "noreply", state
}

// Terminate called when process died
func (dgs *demoGenServ) Terminate(reason string, state interface{}) {
	fmt.Printf("Terminate: %#v\n", reason)
}

func init() {
	flag.IntVar(&ListenRangeBegin, "listen_begin", 15151, "listen port range")
	flag.IntVar(&ListenRangeEnd, "listen_end", 25151, "listen port range")
	flag.StringVar(&GenServerName, "gen_server_name", "example", "gen_server name")
	flag.StringVar(&NodeName, "name", "demo@127.0.0.1", "node name")
	flag.IntVar(&ListenEPMD, "epmd", 4369, "EPMD port")
	flag.StringVar(&Cookie, "cookie", "123", "cookie for interaction with erlang cluster")
}

func main() {
	flag.Parse()

	opts := ergo.NodeOptions{
		ListenRangeBegin: uint16(ListenRangeBegin),
		ListenRangeEnd:   uint16(ListenRangeEnd),
		EPMDPort:         uint16(ListenEPMD),
	}

	// Initialize new node with given name, cookie, listening port range and epmd port
	node := ergo.CreateNode(NodeName, Cookie, opts)

	// Initialize new instance of demoGenServ structure which implements Process behaviour
	demoGS := new(demoGenServ)

	// Spawn process with one arguments
	process, _ := node.Spawn(GenServerName, ergo.ProcessOptions{}, demoGS)
	fmt.Println("Run erl shell:")
	fmt.Printf("erl -name %s -setcookie %s\n", "erl-"+node.FullName, Cookie)

	fmt.Println("-----Examples that can be tried from 'erl'-shell")
	fmt.Printf("gen_server:cast({%s,'%s'}, stop).\n", GenServerName, NodeName)
	fmt.Printf("gen_server:call({%s,'%s'}, hello).\n", GenServerName, NodeName)

	process.Wait()
	node.Stop()
	node.Wait()
}
```

Then `gen_server:call({%s,'%s'}, time).`

**Comment by heri16** [comment]
See my playground: https://play.golang.org/p/5mgvqLnzgsy

**Comment by heri16** [comment]
This is my current workaround, but is problematic for deeply nested values in a `[]map[string]interface{}`.

Would still prefer if ergo does not panic on deeply nested values like `time.Time`.

```go
func GetUnexportedField(field reflect.Value) interface{} {
	return reflect.NewAt(field.Type(), unsafe.Pointer(field.UnsafeAddr())).Elem().Interface()
}

func MarshalText(value interface{}) ([]byte, error) {
	switch v := value.(type) {
	case string:
		return []byte(v), nil
	case encoding.TextMarshaler:
		return v.MarshalText()
	case json.Marshaler:
		return v.MarshalJSON()
	case reflect.Value:
		if v.CanInterface() {
			return MarshalText(v.Interface())
		} else {
			return MarshalText(GetUnexportedField(v))
		}
	case fmt.Stringer:
		return []byte(v.String()), nil
	default:
		return nil, fmt.Errorf("unexpected type: %T value: '%v'", v, v)
	}
}
```

**Comment by halturin** [comment]
Nice catch. Thanks. Just merged PR #64 to fix this issue.

**Comment by halturin** [comment]
released new tag version v.1.2.5
